### PR TITLE
Updates the resource API by adding Resource Categories Mapping

### PR DESCRIPTION
**Pull request by PuneetPunamiya**
Adds a relation between Resources and Categories
    
- This patch updates the design of Resources by adding
      an array of Categories for the following API's
    
         - /query
         - /resources
         - /resource/<resourceID>
         - /v1/query
         - /v1/resources
         - /v1/resource/<resourceID>

 - Updates the model of Category and Resource by creating
   many to many relation
    
 - Adds a migration to update the relation between category
   and resources
    
 - Updates the resources tests which adds category array in
   the response of query, list and resource by id API's
    
 Signed-off-by: Puneet Punamiya <ppunamiy@redhat.com>

# Changes

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [ ] Includes [tests](https://github.com/tektoncd/community/blob/main/standards.md#principles) (if functionality changed/added)
- [ ] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/main/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/pipeline/blob/main/CONTRIBUTING.md) for more details._

### CODE REVIEW

File: api/pkg/db/migration/202106031300_update_resources_category_table.go [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+package migration
+
+import (
+	"github.com/go-gormigrate/gormigrate/v2"
+	"github.com/tektoncd/hub/api/gen/log"
+	"github.com/tektoncd/hub/api/pkg/db/model"
+	"gorm.io/gorm"
+)
+
+func updateResourcesCategoryTable(log *log.Logger) *gormigrate.Migration {
```
Comment (sm43) [inline api/pkg/db/migration/202106031300_update_resources_category_table.go:24]: can you mention what change will this have ?
in the name or a comment would be nice

File: api/pkg/db/model/model.go [hunk 2]
```diff
@@ -59,6 +60,7 @@ type (
 		Kind      string `gorm:"not null;default:null"`
 		Rating    float64
 		Catalog   Catalog
+		Category  []*Category `gorm:"many2many:resource_categories;"`
```
Comment (sm43) [inline api/pkg/db/model/model.go:63]: `Categories`?

File: api/pkg/service/catalog/syncer.go [hunk 3]
```diff
@@ -255,6 +255,7 @@ func (s *syncer) updateResources(
 
 		log.Infof("Resource: %s  ID: %d stored", r.Name, dbRes.ID)
 
+		s.updateResourceCategory(txn, log, &dbRes, r.Categories)
```
Comment (pratap0007) [inline api/pkg/service/catalog/syncer.go:258]: can we pass resource id here instead of `dbRes`

**Comment by pratap0007** [comment]
Have added a comment and it looks good to me

File: api/pkg/service/catalog/syncer.go (hunk 3 above)
Comment (PuneetPunamiya) [reply api/pkg/service/catalog/syncer.go:258]: I think in order to maintain consistency wrt to other methods passed the resource object
Shall I still pass the id ?

File: api/v1/service/resource/resource_http_test.go [hunk 4]
```diff
@@ -176,11 +176,96 @@ func TestQueryWithKindsAndTags_Http(t *testing.T) {
 	})
 }
 
+// func TestQueryWithAllParams_Http(t *testing.T) {
```
Comment (sm43) [inline api/v1/service/resource/resource_http_test.go:179]: wanna remove?

File: api/pkg/parser/catalog.go [hunk 5]
```diff
@@ -41,14 +41,16 @@ const (
 	DisplayNameAnnotation         = "tekton.dev/displayName"
 	MinPipelinesVersionAnnotation = "tekton.dev/pipelines.minVersion"
 	TagsAnnotation                = "tekton.dev/tags"
+	CategoryAnnotation            = "tekton.dev/category"
```
Comment (sm43) [inline api/pkg/parser/catalog.go:44]: category or categories?

**Review by sm43 — CHANGES_REQUESTED** [review]

File: api/pkg/service/catalog/syncer.go (hunk 3 above)
Comment (pratap0007) [reply api/pkg/service/catalog/syncer.go:258]: no, okay then use the resource object

File: api/pkg/parser/catalog.go (hunk 5 above)
Comment (PuneetPunamiya) [reply api/pkg/parser/catalog.go]: Updated !

File: api/pkg/shared/resource/resource.go [hunk 6]
```diff
@@ -66,6 +67,7 @@ func (r *Request) Query() ([]model.Resource, error) {
 		filterByTags(r.Tags),
 		filterByKinds(r.Kinds),
 		filterByCatalogs(r.Catalogs),
+		filterByCategories(r.Categories),
```
Comment (sm43) [inline api/pkg/shared/resource/resource.go:70]: `// Fields: name, []kinds,[]Catalogs []Categories []Tags, Match, Limit`

File: api/test/fixtures/resource_categories.yaml [hunk 7]
```diff
@@ -0,0 +1,31 @@
+# Copyright © 2020 The Tekton Authors.
```
Comment (sm43) [inline api/test/fixtures/resource_categories.yaml:1]: 2021

File: api/pkg/shared/resource/resource.go (hunk 6 above)
Comment (PuneetPunamiya) [reply api/pkg/shared/resource/resource.go:70]: Should this be added as a comment ??

Comment (sm43) [reply api/pkg/shared/resource/resource.go:70]: comment is the just update with ` []Categories `

Comment (sm43) [reply api/pkg/shared/resource/resource.go:70]: line no 51

**Pull request merged**